Zend/tests/readonly_props/readonly_clone.phpt

```diff
@@ -0,0 +1,63 @@
+--TEST--
+clone can write to readonly properties
+--FILE--
+<?php
+
+class Counter
+{
+    private static int $counter = 0;
+
+    public readonly int $count;
+
+    public function __construct()
+    {
+        $this->count = ++self::$counter;
+    }
+
+    public function count(?int $count = null): static
+    {
+        $new = clone $this;
+        $new->count = $count ?? ++self::$counter;
+
+        return $new;
+    }
+
+    public function __clone()
+    {
+        if (is_a(debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 2)[1]['class'] ?? '', self::class, true)) {
```

Comment:
This check is required to allow wither methods (count in this case) to further initialize the cloned object.
I'd be happy to gather ideas to make this either easier or unnecessary.

Reply:
Here is an idea:

```php
function withFoo($newFoo)
{
	// This syntax would be allowed only inside methods and only to clone `$this`
	// in the body of the `clone` scope, `$this` is already the cloned instance
	return clone {
		$this->foo = $newFoo;
	};
}
```

Reply:
Here is another idea:

```php
class Abc
{
    // The `clone` modifier declares that $this inside the method should be a cloned instance.
    // Such "clone methods" would be allowed to return only `static` and would be exposed
    // by reflection via eg `ReflectionMethod::isClone()`
    public clone function withFoo($newFoo): static
    {
        $this->foo = $newFoo;

        return $this;
    }
}
```

Reply:
alternative without new `clone function` keyword:
`clone($obj, 'foo', 'bar');`
`public function __clone($arg1, $arg2);`

controlling "internal call only" and enforcing withers can be done with visibility: `private function __clone(...$args)`

Reply:
This concern (withers) is purely internal. I don't think the syntax we're looking for should force exposing any public API (arguments in your proposal).
`private function __clone(...)` would mean consumers can't clone (at least that'd be the expected behavior to many since the visibility modifier means that - visibility to consumers.)

Reply:
Also: the benefit of the `clone` modifier is that it defines a very clear scope where updating readonly propoerties is safe and thus allowed: the body of the method. With a syntax like `clone($foo)`, we don't know have such "init-safe" scope.

Reply:
Just a note, there is copious previous comparison of different options for cloning in my blog posts on the topic:

https://peakd.com/hive-168588/@crell/object-properties-and-immutability
https://peakd.com/hive-168588/@crell/object-properties-part-2-examples

(The first one is long; if you want to skip to the relevant for now part, skip to the "Review" section.)

Reply:
I'll also note that while poking holes in `readonly` is fine, and I'm OK with it, there are still use cases where full asymmetric visibility would be beneficial.

Besides cloning, __unserialize() is another possible place to poke a hole.

Reply:
Happy to contribute another idea to the discussion then @Crell. I'm personally not convinced by the need of a dedicated syntax to set several properties at once. Several statements are fine. But we need a bounded context where re-initializing a readonly prop is allowed. __clone already provides us one, and this `clone` modifier would also. On top, it would simplify writing wither APIs a lot.

But yay, that's a separate discussion from that PR.
This PR doesn't need any new syntax and it makes sense on its own.
I opened this thread, but I'm closing it as this is not the right place to discuss this :)
Thanks for coming by.